mls-rs-uniffi/src/lib.rs

```diff
@@ -0,0 +1,528 @@
+    pub fn new(id: Vec<u8>, signature_keypair: SignatureKeypair) -> Self {
+        let cipher_suite = signature_keypair.cipher_suite;
+        let public_key = arc_unwrap_or_clone(signature_keypair.public_key);
+        let secret_key = arc_unwrap_or_clone(signature_keypair.secret_key);
+        let crypto_provider = OpensslCryptoProvider::new();
+        let basic_credential = BasicCredential::new(id);
+        let signing_identity =
+            identity::SigningIdentity::new(basic_credential.into_credential(), public_key.inner);
+        Client {
+            inner: mls_rs::Client::builder()
+                .crypto_provider(crypto_provider)
+                .identity_provider(basic::BasicIdentityProvider::new())
+                .signing_identity(signing_identity, secret_key.inner, cipher_suite.into())
+                .build(),
+        }
+    }
+
+    /// Generate a new key package for this client.
+    ///
+    /// The key package is represented in is MLS message form. It is
+    /// needed when joining a group and can be published to a server
+    /// so other clients can look it up.
+    ///
+    /// See [`mls_rs::Client::generate_key_package_message`] for
+    /// details.
+    pub async fn generate_key_package_message(&self) -> Result<Message, Error> {
+        let message = self.inner.generate_key_package_message().await?;
+        Ok(message.into())
+    }
+
+    /// Create and immediately join a new group.
+    ///
+    /// If a group ID is not given, the underlying library will create
+    /// a unique ID for you.
+    ///
+    /// See [`mls_rs::Client::create_group`] and
+    /// [`mls_rs::Client::create_group_with_id`] for details.
+    pub async fn create_group(&self, group_id: Option<Vec<u8>>) -> Result<Group, Error> {
+        let extensions = mls_rs::ExtensionList::new();
+        let inner = match group_id {
+            Some(group_id) => {
+                self.inner
+                    .create_group_with_id(group_id, extensions)
+                    .await?
+            }
+            None => self.inner.create_group(extensions).await?,
+        };
+        Ok(Group {
+            inner: Arc::new(Mutex::new(inner)),
+        })
+    }
+
+    /// Join an existing group.
+    ///
+    /// See [`mls_rs::Client::join_group`] for details.
+    pub async fn join_group(&self, welcome_message: Arc<Message>) -> Result<JoinInfo, Error> {
+        let welcome_message = arc_unwrap_or_clone(welcome_message);
+        let (group, new_member_info) = self.inner.join_group(None, welcome_message.inner).await?;
+
+        let group = Arc::new(Group {
+            inner: Arc::new(Mutex::new(group)),
+        });
+        let group_info_extensions = Arc::new(new_member_info.group_info_extensions.into());
+        Ok(JoinInfo {
+            group,
+            group_info_extensions,
+        })
+    }
+}
+
+#[derive(Clone, Debug, uniffi::Object)]
+pub struct CommitOutput {
+    _inner: mls_rs::group::CommitOutput,
+}
```

Comment:
There seems to be no way to extract commit and welcome messages from this struct?

Reply:
No, it's not hooked up to anything yet — several of them are not used yet. That should definitely change soon!

Reply:
This is now fixed as per @tomleavy's comment above.